### PR TITLE
feat: share kafka client on meta

### PR DESCRIPTION
**Pull request by tabVersion**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

as title, reuse the client if the broker addr is the same, to reduce the conn to the broker

The key changes in this PR revolve around optimizing Kafka client management by introducing connection pooling. Here's the main changes:

Introduction of SHARED_KAFKA_CLIENT:
```
client.rs
// Added a shared cache for Kafka clients
pub static SHARED_KAFKA_CLIENT: LazyLock<MokaCache<KafkaConnection, Weak<KafkaClientType>>> =
    LazyLock::new(|| moka::future::Cache::builder().build());
```

The main motivations appear to be:
1. Resource Optimization: Instead of creating new Kafka client connections for each enumerator, connections are now reused when possible through a shared cache.
2. Memory Management: Uses Weak references to prevent memory leaks (when all related resources are dropped, drop the client then):

```
// Uses Arc and Weak to manage references
type KafkaClientType = BaseConsumer<RwConsumerContext>;
// Cache stores weak references to allow cleanup when clients are no longer needed
MokaCache<KafkaConnection, Weak<KafkaClientType>>
```

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by hzxa21** [comment]
https://github.com/risingwavelabs/risingwave/issues/18949

File: src/connector/src/source/kafka/enumerator/client.rs [hunk 1]
```diff
@@ -32,6 +33,14 @@ use crate::source::kafka::{
 };
 use crate::source::SourceEnumeratorContextRef;
 
+pub static SHARED_KAFKA_CLIENT: LazyLock<tokio::sync::Mutex<HashMap<u64, SharedKafkaItem>>> =
```
Comment (fuyufjh) [inline src/connector/src/source/kafka/enumerator/client.rs:36]:
```suggestion
pub static SHARED_KAFKA_CLIENT: LazyLock<tokio::sync::Mutex<HashMap<KafkaConnection, SharedKafkaItem>>> =
```

Please just use the `KafkaConnection` as the key. The `HashMap` is supposed to compute the hash code and avoid collision.

File: src/connector/src/source/kafka/enumerator/client.rs [hunk 2]
```diff
@@ -94,36 +105,67 @@ impl SplitEnumerator for KafkaSplitEnumerator {
...
+                "reusing kafka client for connection hash {}, to broker {}",
+                connection_hash,
+                broker_address
+            );
+            kafka_client = item.client.clone();
+            item.ref_count += 1;
+            drop(shared_client_guard);
+        } else {
+            // drop the guard and acquire a new one to avoid a 10s blocking call
+            drop(shared_client_guard);
```
Comment (fuyufjh) [inline src/connector/src/source/kafka/enumerator/client.rs:121]: Will this happen?

1. Caller A tries to get a client for a connection, but cache missed, so it takes some time for calller A to build the client.
2. During it, caller B tries to get a client for the same connection, but cache missed, so it also builds a client.
3. Caller A built and inserted the client to the map and set ref_count = 1
4. One second later, caller B also inserted the client to the map and set ref_count = 1, **causing the client of caller A to leak and will never be dropped.**

Comment (fuyufjh) [reply src/connector/src/source/kafka/enumerator/client.rs:121]: I would recommend to use `moka` for this case.

```rust
use moka::future::Cache;
```

Particularly, a caching structure should handle these concurrent gets correctly by letting the caller B blocks until caller A completes its operation and insert back the cached item i.e. the Kafka Client.

File: src/connector/src/sink/kafka.rs [hunk 3]
```diff
@@ -214,6 +214,9 @@ pub struct KafkaConfig {
     #[serde(flatten)]
     pub common: KafkaCommon,
 
+    #[serde(flatten)]
+    pub connection: crate::connector_common::KafkaConnection,
```
Comment (xxchan) [inline src/connector/src/sink/kafka.rs:218]: nit: I guess changing this to `KafkaConnection` can make `src/connector/with_options_sink.yaml` look better

**Review by xxchan — COMMENTED** [review]
Could you please also run some tests and describe the improvements in the PR description?

File: src/connector/src/source/kafka/enumerator/client.rs [hunk 4]
```diff
@@ -94,36 +107,84 @@ impl SplitEnumerator for KafkaSplitEnumerator {
...
+            })
+            .await
+            .into_value();
+
+        let kafka_client: Arc<BaseConsumer<RwConsumerContext>> = match client_arc.as_ref() {
+            Ok(client) => client.client.clone(),
+            Err(e) => {
+                SHARED_KAFKA_CLIENT.remove(&properties.connection).await;
+                #[allow(rw::format_error)]
+                bail!("{}", e.as_report());
```
Comment (tabVersion) [inline src/connector/src/source/kafka/enumerator/client.rs:179]: Here is a lint exception because e is of type `&ConnectorError`, which cannot derive Clone (defined via `def_anyhow_newtype`) and can not `.to_owned`, but return error requires `ConnectorError`. Directly using `as_report` here to keep format.

**Comment by tabVersion** [comment]
> Could you please also run some tests and describe the improvements in the PR description?

I’m not entirely sure what specific tests you’re looking for. This change doesn't impact performance. Running a test that creates 600 sources on CI seems a bit excessive (although we could certainly do it if needed). 
I believe the improvement in reusing the Kafka client on meta is quite clear, and I also think it will help address the issue of too many threads that customers are experiencing.

**Comment by tabVersion** [comment]
Do a quick test with `./risedev d full`

```bash 
for i in {0..100}; do echo "{\"a\": $i}^{\"a\": $i, \"b\": \"$i\"}" | kcat -P -b 127.0.0.1:9092 -t test -K ^ -H "header1=v1" -H "header2=v2"; done

for i in {0..600}; do  psql -h localhost -p 4566 -d dev -U root -c "create source s_$i (a int, b varchar) with (connector = 'kafka', topic = 'test', properties.bootstrap.server = '127.0.0.1:29092') format plain encode json ;" ; done
```

after creating over 120 sources, got the error:

```text
Caused by these errors (recent errors listed first):
  1: gRPC request to meta service failed: Internal error
  2: The cluster is recovering
  3: get error from control stream, in worker node 1
  4: gRPC request to stream service failed: Internal error
  5: recv actor failure
  6: Actor 506 exited unexpectedly
  7: Executor error
  8: Connector error
  9: Kafka error
 10: Meta data fetch error
 11: AllBrokersDown (Local: All broker connections are down)
```
which indicates kafka broker is down and meta is okay.

**Comment by xxchan** [comment]
> I’m not entirely sure what specific tests you’re looking for.

At least sth like "Manually tested that num of threads is reduced from xxx to yyy for zzz Kafka sources."

Although the idea might be clear, the implementation is not that trivial, so we should verify it works.

Besides, the background of the problem should also be mentioned.

File: Cargo.lock [hunk 5]
```diff
@@ -16613,3 +16530,8 @@ dependencies = [
  "libc",
  "pkg-config",
 ]
+
+[[patch.unused]]
+name = "quanta"
+version = "0.11.0"
+source = "git+https://github.com/madsim-rs/quanta.git?rev=948bdc3#948bdc3d4cd3fcfe3d52d03dd83deee96d97d770"
```
Comment (xxchan) [inline Cargo.lock:16537]: I guess this might break madsim

File: src/connector/src/connector_common/common.rs [hunk 6]
```diff
@@ -161,21 +162,11 @@ impl AwsAuthProps {
 }
 
 #[serde_as]
-#[derive(Debug, Clone, Deserialize, WithOptions)]
-pub struct KafkaCommon {
+#[derive(Debug, Clone, Deserialize, WithOptions, PartialEq, Hash, Eq)]
+pub struct KafkaConnection {
```
Comment (xxchan) [inline src/connector/src/connector_common/common.rs:166]: It seems we will be able to use this same struct for `CREATE CONNECTION` later? 🤔

File: src/connector/src/connector_common/common.rs [hunk 7]
```diff
@@ -316,7 +321,13 @@ impl RdKafkaPropertiesCommon {
...
 }
 
-impl KafkaCommon {
+impl KafkaConnection {
+    pub fn get_hash(&self) -> u64 {
+        let mut hasher = std::collections::hash_map::DefaultHasher::new();
+        self.hash(&mut hasher);
+        hasher.finish()
+    }
+
```
Comment (xxchan) [inline src/connector/src/connector_common/common.rs:330]:
```suggestion
```

File: src/connector/src/connector_common/common.rs (hunk 6 above)
Comment (tabVersion) [reply src/connector/src/connector_common/common.rs:166]: yes.

File: src/meta/src/stream/source_manager.rs [hunk 8]
```diff
@@ -1033,6 +1038,46 @@ impl SourceManager {
...
+                if let Some(entry) = handle.share_client_entry {
+                    SHARED_KAFKA_CLIENT
+                        .entry_by_ref(&entry)
+                        .and_compute_with(|maybe_item| async {
+                            if let Some(item) = maybe_item {
+                                match item.into_value().as_ref() {
+                                    Ok(item_val) => {
+                                        let share_item = SharedKafkaItem {
+                                            client: item_val.client.clone(),
+                                            ref_count: item_val.ref_count - 1,
```
Comment (xxchan) [inline src/meta/src/stream/source_manager.rs:1051]: I don't get why do we need `ref_count`. Can we just use sth like `get_with`?

Comment (tabVersion) [reply src/meta/src/stream/source_manager.rs:1051]: Then how do we manage the RdKafka client instance if all related sources are dropped? IIUC, if we remove the ref_count, the client instance will always be kept in memory, until restarting the meta node. There is no evict policy enabled for the cache.

File: src/meta/src/stream/source_manager.rs [hunk 9]
```diff
@@ -1033,6 +1038,46 @@ impl SourceManager {
         for source_id in source_ids {
             if let Some(handle) = core.managed_sources.remove(&source_id) {
                 handle.handle.abort();
+                // remove the shared kafka client entry if it exists
+                if let Some(entry) = handle.share_client_entry {
```
Comment (BugenZhao) [inline src/meta/src/stream/source_manager.rs:1042]: Is there any reason for manually maintaining the ref count? Does simply storing a `Weak` in the cache work?

File: src/meta/src/stream/source_manager.rs (hunk 8 above)
Comment (xxchan) [reply src/meta/src/stream/source_manager.rs:1051]: Oh... I'm a little confused now. Then should we use a concurrent hashmap like `dashmap`, instead of a cache..?

File: src/connector/src/source/kafka/enumerator/client.rs [hunk 10]
```diff
@@ -13,25 +13,38 @@
...
 use crate::error::ConnectorResult;
 use crate::source::base::SplitEnumerator;
 use crate::source::kafka::split::KafkaSplit;
 use crate::source::kafka::{
-    KafkaContextCommon, KafkaProperties, RwConsumerContext, KAFKA_ISOLATION_LEVEL,
+    KafkaConnection, KafkaContextCommon, KafkaProperties, RwConsumerContext, KAFKA_ISOLATION_LEVEL,
 };
 use crate::source::SourceEnumeratorContextRef;
 
+pub static SHARED_KAFKA_CLIENT: LazyLock<
```
[the remark this answers is not in the document]
Comment (tabVersion) [reply src/connector/src/source/kafka/enumerator/client.rs:38]: Oh, this one is what I am looking for. Let me fix it.

File: src/meta/src/stream/source_manager.rs (hunk 8 above)
Comment (xxchan) [reply src/meta/src/stream/source_manager.rs:1051]: As mentioned by Bugen, I think we can store `Weak` in the cache.

File: Cargo.lock (hunk 5 above)
Comment (tabVersion) [reply Cargo.lock:16537]: Is there a way to fix it? Directly removing the part does not help.

File: src/meta/src/stream/source_manager.rs (hunk 8 above)
Comment (BugenZhao) [reply src/meta/src/stream/source_manager.rs:1051]: Not sure if `Weak` fits well with `moka`. We may also try `dashmap` or `weak-table`.

**Comment by tabVersion** [comment]
> > I’m not entirely sure what specific tests you’re looking for.
> 
> At least sth like "Manually tested that num of threads is reduced from xxx to yyy for zzz Kafka sources."
> 
> Although the idea might be clear, the implementation is not that trivial, so we should verify it works.
> 
> Besides, the background of the problem should also be mentioned.

Tested locally

Kafka env

```
Metadata for test (from broker 0: 127.0.0.1:9092/0):
 1 brokers:
  broker 0 at 127.0.0.1:9092 (controller)
 1 topics:
  topic "test" with 1 partitions:
    partition 0, leader 0, replicas: 0, isrs: 0
```

command: 

```
for i in {0..100}; do  psql -h localhost -p 4566 -d dev -U root -c "create source s_$i (a int, b varchar) with (connector = 'kafka', topic = 'test', properties.bootstrap.server = '127.0.0.1:9092') format plain encode json ;" ; done
```

on main (a176aced3c047d780adc8f3981a4921faa526255): 1573 threads
on this pr (ac1d63d28f7ac5441d2260324184d13acc0d049f): 1272 threads

File: Cargo.lock (hunk 5 above)
Comment (tabVersion) [reply Cargo.lock:16537]: after removing 
```toml
# from Cargo.toml, in `[patch.crates-io]`
quanta = { git = "https://github.com/madsim-rs/quanta.git", rev = "948bdc3" }
```
the `[[patch.unused]]` section disappears. But the tests still stuck

```
  Running [ 00:01:43] [=============================================================================>                                                                                            ]  533/1167: 12 running, 533 passed (4 slow), 22 skipped     
   Canceling due to interrupt: 12 tests still running
   Canceling [ 00:01:43] [=============================================================================>                                                                                            ]  533/1167: 12 running, 533 passed (4 slow), 22 skipped     
      SIGINT [ 103.327s] risingwave_batch executor::generic_exchange::tests::test_exchange_multiple_sources
      SIGINT [ 102.891s] risingwave_batch executor::merge_sort_exchange::tests::test_exchange_multiple_sources
      SIGINT [ 102.802s] risingwave_batch executor::tests::test_clone_for_plan
      SIGINT [ 102.726s] risingwave_batch task::task_manager::tests::test_task_id_conflict
      SIGINT [  83.848s] risingwave_frontend binder::bind_param::test::subquery
      SIGINT [  83.893s] risingwave_frontend binder::bind_param::test::cast_after_specific
      SIGINT [  83.899s] risingwave_frontend binder::bind_param::test::basic_value
      SIGINT [  83.887s] risingwave_frontend binder::bind_param::test::default_type
      SIGINT [  83.903s] risingwave_frontend binder::bind_param::test::basic_select
      SIGINT [  83.042s] risingwave_frontend binder::expr::value::tests::test_bind_radix
      SIGINT [  83.886s] risingwave_frontend binder::bind_param::test::infer_case
      SIGINT [  83.498s] risingwave_frontend binder::expr::value::tests::test_bind_interval
------------
```

the dep to `quanta` is introduced by `moka`

Comment (xxchan) [reply Cargo.lock:16537]: Patching `quanta` is **required**, not unexpected.

moka 0.12.0 depends on quanta 0.11, but you upgraded moka to 0.12.8, which depends on quanta 0.12. Therefore, original quanta is used instead of the patched one. That's why it breaks.

image

Comment (xxchan) [reply Cargo.lock:16537]: You should either revert to moka 0.12.0, or upgrade the madsim quanta to 0.12

File: src/meta/src/stream/source_manager.rs (hunk 8 above)
Comment (xxchan) [reply src/meta/src/stream/source_manager.rs]: @BugenZhao Why do you think `Weak` might not work with `moka`? Actually I'm also thinking what's the difference between `moka` (cache) and `dashmap` (concurrent hashmap)

**Comment by xxchan** [comment]
Would you mind testing Kafka with multiple brokers, where we might see a larger difference?

**Comment by tabVersion** [comment]
kafka cluster info ([docker compose](https://docs.redpanda.com/redpanda-labs/docker-compose/_attachments/three-brokers/docker-compose.yml))

```
Metadata for all topics (from broker -1: 127.0.0.1:19092/bootstrap):
 3 brokers:
  broker 1 at localhost:29092
  broker 2 at localhost:39092
  broker 0 at localhost:19092 (controller)
 5 topics:
  topic "_schemas" with 1 partitions:
    partition 0, leader 0, replicas: 1,0,2, isrs: 1,0,2
  topic "test" with 1 partitions:
    partition 0, leader 2, replicas: 2, isrs: 2
  topic "topic1" with 3 partitions:
    partition 0, leader 0, replicas: 1,2,0, isrs: 1,2,0
    partition 1, leader 1, replicas: 1,0,2, isrs: 1,0,2
    partition 2, leader 2, replicas: 1,0,2, isrs: 1,0,2
  topic "topic2" with 3 partitions:
    partition 0, leader 1, replicas: 1,2,0, isrs: 1,2,0
    partition 1, leader 2, replicas: 1,0,2, isrs: 1,0,2
    partition 2, leader 0, replicas: 0,1,2, isrs: 0,1,2
  topic "topic3" with 3 partitions:
    partition 0, leader 1, replicas: 2,1,0, isrs: 2,1,0
    partition 1, leader 2, replicas: 1,2,0, isrs: 1,2,0
    partition 2, leader 0, replicas: 0,1,2, isrs: 0,1,2
```

topic1, topic2, topic3 are all of 3 partition and 3 replica

```
declare -a topics=("topic1" "topic2" "topic3")
random_topic=${topics[$RANDOM % ${#topics[@]}]}

for i in {0..100}; do psql -h localhost -p 4566 -d dev -U root -c "create table s_${RANDOM} (a int, b varchar) with (connector = 'kafka', topic = '$random_topic', properties.bootstrap.server = 'localhost:19092,localhost:39092,localhost:29092') format plain encode json;" ; done 
```

got the error after creating 20+ tables, will investigate later

```
ERROR:  Failed to run the query

Caused by these errors (recent errors listed first):
  1: gRPC request to meta service failed: Internal error
  2: get error from control stream, in worker node 1
  3: gRPC request to stream service failed: Internal error
  4: recv actor failure
  5: Actor 312 exited unexpectedly
  6: Executor error
  7: Connector error
  8: Kafka error
  9: Meta data fetch error
 10: Resolve (Local: Host resolution failure)
```

File: src/meta/src/stream/source_manager.rs (hunk 8 above)
Comment (tabVersion) [reply src/meta/src/stream/source_manager.rs]: > Actually I'm also thinking what's the difference between moka (cache) and dashmap (concurrent hashmap)

I think moka is a dashmap with evict policy and guarantees the updates can be done atomically.

Comment (BugenZhao) [reply src/meta/src/stream/source_manager.rs]: I think `dashmap` behaves similar to `moka::sync::Cache` with unlimited capacity. 🤔

> Why do you think `Weak` might not work with `moka`?

Because the interface does not seem to be _that_ compatible with storing a `Weak`, like, no auto-eviction, requires the caller to keep the strong reference:

```rust
 let mut client_arc: Option<Arc<KafkaClientType>> = None;
```

Comment (xxchan) [reply src/meta/src/stream/source_manager.rs]: > requires the caller to keep the strong reference

Isn't this expected usage? i.e., store `Weak` in the map, while store `Arc` in the caller. Otherwise who keeps the `Arc`? :eyes:

Comment (BugenZhao) [reply src/meta/src/stream/source_manager.rs]: I mean when inserting. 🤣 For example, `weak-table` allows you to pass a closure returning an `Arc` in `insert_with` while actually storing a `Weak` but returning an `Arc` back to the caller. With `moka` we need to temporarily hold the strong reference to prevent it from being deallocated.

**Comment by tabVersion** [comment]
> Would you mind testing Kafka with multiple brokers, where we might see a larger difference?

tested with confluent with multiple AZ

```
Metadata for all topics (from broker -1: sasl_ssl://pkc-p11xm.us-east-1.aws.confluent.cloud:9092/bootstrap):
 18 brokers:
  broker 0 at b0-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
  broker 1 at b1-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
  broker 2 at b2-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
  broker 3 at b3-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
  broker 4 at b4-pkc-p11xm.us-east-1.aws.confluent.cloud:9092 (controller)
  broker 5 at b5-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
  broker 6 at b6-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
  broker 7 at b7-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
  broker 8 at b8-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
  broker 9 at b9-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
  broker 10 at b10-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
  broker 11 at b11-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
  broker 12 at b12-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
  broker 13 at b13-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
  broker 14 at b14-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
  broker 15 at b15-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
  broker 16 at b16-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
  broker 17 at b17-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
 1 topics:
  topic "topic_0" with 6 partitions:
    partition 0, leader 12, replicas: 12,17,13, isrs: 12,17,13
    partition 1, leader 17, replicas: 17,13,15, isrs: 17,13,15
    partition 2, leader 13, replicas: 13,15,14, isrs: 13,15,14
    partition 3, leader 15, replicas: 15,14,16, isrs: 15,14,16
    partition 4, leader 14, replicas: 14,16,12, isrs: 14,16,12
    partition 5, leader 16, replicas: 16,12,17, isrs: 16,12,17
```

testing script 

```
for i in {0..4}; do 
	psql  -h 127.0.0.1 -p 4566 -d dev -U root -c "create source s_$i(a int, b varchar) with (connector = 'kafka', topic = 'topic_0', properties.bootstrap.server = 'xxx.us-east-1.aws.confluent.cloud:9092', properties.security.protocol = 'SASL_SSL', properties.sasl.mechanism = 'PLAIN', properties.sasl.username = '[...]', properties.sasl.password = '[...]') format plain encode json;" ; 
done
```

EC2 idle: 71
EC2 with risingwave running: 195
on main (https://github.com/risingwavelabs/risingwave/commit/a176aced3c047d780adc8f3981a4921faa526255): 744
on this pr (https://github.com/risingwavelabs/risingwave/commit/ac1d63d28f7ac5441d2260324184d13acc0d049f): 654

---

why not test with more source?

creating 5 source at once seems the SDK's maximum, I got the error afterward:

```
Caused by these errors (recent errors listed first):         
  1: gRPC request to meta service failed: Internal error
  2: The cluster is recovering                               
  3: get error from control stream, in worker node 1
  4: gRPC request to stream service failed: Internal error
  5: recv actor failure                                                                                                                                                                                                                              
  6: Actor 175 exited unexpectedly                 
  7: Executor error                            
  8: Connector error                                                                                                      
  9: Kafka error                                                                                                          
 10: Meta data fetch error                                                                                                                                                                                                                           
 11: Resolve (Local: Host resolution failure)
```

a little weird but irrelevant to the issue.

File: src/meta/Cargo.toml
Comment (BugenZhao) [inline src/meta/Cargo.toml:1]: Can revert changes in this file?

File: src/connector/Cargo.toml [hunk 11]
```diff
@@ -71,7 +71,7 @@ jni = { version = "0.21.1", features = ["invocation"] }
 jsonbb = { workspace = true }
 jsonwebtoken = "9.2.0"
 maplit = "1.0.2"
-moka = { version = "0.12.0", features = ["future"] }
+moka = { version = "0.12.8", features = ["future"] }
```
Comment (BugenZhao) [inline src/connector/Cargo.toml:74]: Can we still use `0.12.0` in this PR, so that there's no changes in `Cargo.lock`and bumping of `quanta` can be reviewed in a separate PR?

**Review by BugenZhao — COMMENTED** [review]
Rest LGTM

File: src/connector/Cargo.toml (hunk 11 above)
Comment (tabVersion) [reply src/connector/Cargo.toml:74]: I am afraid not. We are using `and_try_compute_with`, which is unavailable in v0.12.0.

Comment (BugenZhao) [reply src/connector/Cargo.toml:74]: Would you try `0.12.3`?

Comment (tabVersion) [reply src/connector/Cargo.toml:74]: I think it works

Comment (BugenZhao) [reply src/connector/Cargo.toml:74]: You may want to checkout the `Cargo.lock` on the main branch to downgrade the version of `quanta`.

Comment (tabVersion) [reply src/connector/Cargo.toml:74]: image

moka 0.12.3 still needs a higher version of `quanta`. I am afraid we have to bump quanta first.

Comment (BugenZhao) [reply src/connector/Cargo.toml:74]: You're right. 😢 In its repo the version `0.12.3` released before bumping dependency on `quanta`, but on crates.io it's the other way around.

File: Cargo.toml [hunk 12]
```diff
@@ -343,7 +339,7 @@ opt-level = 2
 
 [patch.crates-io]
 # Patch third-party crates for deterministic simulation.
-quanta = { git = "https://github.com/madsim-rs/quanta.git", rev = "948bdc3" }
+quanta = { git = "https://github.com/tabVersion/quanta.git", rev = "bb6c780894d06c0ec3f487d58c72920665b5cb0a" }
```
Comment (BugenZhao) [inline Cargo.toml:342]: We may contribute to `madsim-rs/quanta`.

**Review by BugenZhao — APPROVED** [review]

File: Cargo.toml (hunk 12 above)
Comment (tabVersion) [reply Cargo.toml:342]: let's merge https://github.com/madsim-rs/quanta/pull/2 first and we can switch back to madsim.

**Comment by xxchan** [comment]
Do we want to cherry-pick?

**Comment by lmatz** [comment]
> > Would you mind testing Kafka with multiple brokers, where we might see a larger difference?
> 
> tested with confluent with multiple AZ
> 
> ```
> Metadata for all topics (from broker -1: sasl_ssl://pkc-p11xm.us-east-1.aws.confluent.cloud:9092/bootstrap):
>  18 brokers:
>   broker 0 at b0-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
>   broker 1 at b1-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
>   broker 2 at b2-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
>   broker 3 at b3-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
>   broker 4 at b4-pkc-p11xm.us-east-1.aws.confluent.cloud:9092 (controller)
>   broker 5 at b5-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
>   broker 6 at b6-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
>   broker 7 at b7-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
>   broker 8 at b8-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
>   broker 9 at b9-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
>   broker 10 at b10-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
>   broker 11 at b11-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
>   broker 12 at b12-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
>   broker 13 at b13-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
>   broker 14 at b14-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
>   broker 15 at b15-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
>   broker 16 at b16-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
>   broker 17 at b17-pkc-p11xm.us-east-1.aws.confluent.cloud:9092
>  1 topics:
>   topic "topic_0" with 6 partitions:
>     partition 0, leader 12, replicas: 12,17,13, isrs: 12,17,13
>     partition 1, leader 17, replicas: 17,13,15, isrs: 17,13,15
>     partition 2, leader 13, replicas: 13,15,14, isrs: 13,15,14
>     partition 3, leader 15, replicas: 15,14,16, isrs: 15,14,16
>     partition 4, leader 14, replicas: 14,16,12, isrs: 14,16,12
>     partition 5, leader 16, replicas: 16,12,17, isrs: 16,12,17
> ```
> 
> testing script
> 
> ```
> for i in {0..4}; do 
> 	psql  -h 127.0.0.1 -p 4566 -d dev -U root -c "create source s_$i(a int, b varchar) with (connector = 'kafka', topic = 'topic_0', properties.bootstrap.server = 'xxx.us-east-1.aws.confluent.cloud:9092', properties.security.protocol = 'SASL_SSL', properties.sasl.mechanism = 'PLAIN', properties.sasl.username = '[...]', properties.sasl.password = '[...]') format plain encode json;" ; 
> done
> ```
> 
> EC2 idle: 71 EC2 with risingwave running: 195 on main ([a176ace](https://github.com/risingwavelabs/risingwave/commit/a176aced3c047d780adc8f3981a4921faa526255)): 744 on this pr ([ac1d63d](https://github.com/risingwavelabs/risingwave/commit/ac1d63d28f7ac5441d2260324184d13acc0d049f)): 654
> 
> why not test with more source?
> 
> creating 5 source at once seems the SDK's maximum, I got the error afterward:
> 
> ```
> Caused by these errors (recent errors listed first):         
>   1: gRPC request to meta service failed: Internal error
>   2: The cluster is recovering                               
>   3: get error from control stream, in worker node 1
>   4: gRPC request to stream service failed: Internal error
>   5: recv actor failure                                                                                                                                                                                                                              
>   6: Actor 175 exited unexpectedly                 
>   7: Executor error                            
>   8: Connector error                                                                                                      
>   9: Kafka error                                                                                                          
>  10: Meta data fetch error                                                                                                                                                                                                                           
>  11: Resolve (Local: Host resolution failure)
> ```
> 
> a little weird but irrelevant to the issue.

Does the same error occur before this PR under the same setup?
